src/test/scala/is/hail/io/ImportMatrixSuite.scala

```diff
@@ -6,62 +6,118 @@ import is.hail.check.Gen
+      val vsm = LoadMatrix(hc, files, Some(Array("v")), Seq(TString()), Some("v"))
     }
     assert(e.getMessage.contains("invalid sample ids"))
   }
 
   @Test def testMissingVals() {
     val files = hc.hadoopConf.globAll(List("src/test/resources/samplesmissing.txt"))
     val e = intercept[SparkException] {
-      val vsm = LoadMatrix(hc, files, Some(Array("v")), Array(TString()), "v")
+      val vsm = LoadMatrix(hc, files, Some(Array("v")), Seq(TString()), Some("v"))
       vsm.rdd2.count()
     }
     assert(e.getMessage.contains("number of elements"))
   }
 
   @Test def testTypes() {
-    val genMatrix = VSMSubgen(
-      sSigGen = Gen.const(TString()),
-      saSigGen = Gen.const(TStruct.empty()),
-      vSigGen = Gen.const(TString()),
-      vaSigGen = Gen.const(TStruct.empty()),
-      globalSigGen = Gen.const(TStruct.empty()),
-      tSigGen = Gen.zip(Gen.oneOf[Type](TInt32(), TInt64(), TFloat32(), TFloat64(), TString()), Gen.coin(0.2))
-        .map { case (typ, req) => typ.setRequired(req) }.map { t => TStruct("x" -> t) },
-      sGen = (t: Type) => Gen.identifier.map(s => s: Annotation),
-      saGen = (t: Type) => t.genNonmissingValue,
-      vaGen = (t: Type) => t.genNonmissingValue,
-      globalGen = (t: Type) => t.genNonmissingValue,
-      vGen = (t: Type) => Gen.identifier,
-      tGen = (t: Type, v: Annotation) => t.genValue)
-
     forAll(MatrixTable.gen(hc, genMatrix)
       .filter(vsm => !vsm.sampleIds.contains("v"))) { vsm =>
       val actual: MatrixTable = {
         val f = tmpDir.createTempFile(extension = "txt")
         vsm.makeKT("v = v", "`` = g.x", Array("v")).export(f)
-        LoadMatrix(hc, Array(f), None, Array(TString()), "v", cellType = vsm.genotypeSignature)
+        LoadMatrix(hc, Array(f), None, Seq(TString()), Some("v"), cellType = vsm.genotypeSignature)
       }
       assert(vsm.same(actual.annotateVariantsExpr("va = {}")))
 
+
       val tmp1 = tmpDir.createTempFile(extension = "vds")
       vsm.write(tmp1, true)
 
       val vsm2 = MatrixTable.read(hc, tmp1)
       assert(vsm.same(vsm2))
       true
     }.check()
+  }
+
+  @Test def testTypesNoHeader() {
+    forAll(MatrixTable.gen(hc, genMatrix)
+      .filter(vsm => !vsm.sampleIds.contains("v"))) { vsm =>
+
+      val actual2: MatrixTable = {
+        val f = tmpDir.createTempFile(extension = "txt")
+        vsm.makeKT("v = v", "`` = g.x", Array("v")).export(f, header=false)
+        println(f)
+        LoadMatrix(hc, Array(f), Some(Array("v")), Seq(TString()), Some("v"), cellType = vsm.genotypeSignature, noHeader=true)
+      }
+      assert(vsm.copy2(sampleIds = actual2.sampleIds).same(actual2.annotateVariantsExpr("va = {}")), "not same")
+
+
+      val tmp1 = tmpDir.createTempFile(extension = "vds")
+      vsm.write(tmp1, true)
 
+      val vsm2 = MatrixTable.read(hc, tmp1)
+      assert(vsm.same(vsm2), "not same")
+      true
+    }.check()
+  }
+
+  @Test def testTypesNoKey() {
```

Comment:
Can we add a test for the case where there are `annotationHeader == None` but we supply a key expression in terms of the `fi`s?

Reply:
And how about some cases where `noHeader == false` but `annotationHeader == Some(...)`

Reply:
`noHeader == false` and `annotationHeader == Some(...)` is kind of difficult because I couldn't find a natural way to export something as a tsv without all the header fields, but I added the other case.